### PR TITLE
[Notifier] Add `from` in `SmsMessage`

### LINKED ISSUE #45435: [Notifier/SmsMessage] Add setFrom method
### Description

We would love to start using Symfony Notifier for sending SMS notifications in our project. We have just one blocker. 
In most countries, we use our brand name in the sender name. But due to regulations in the U.S., we have to use a regular number. Due to the volume of messages we send, we have several numbers that we use evenly.

Right now we can define default 'from' when we set DNS ([Docs](https://symfony.com/doc/current/notifier.html#sms-channel)). But there is no posibility to change this default variable.

What do you think about adding `setForm` method on `SmsMessage` class? I think this optional step will be really helpful not only for us :)
If you agree I can prepare PR - it looks like not really complicated issue :)

We love what are you doing :) 
Greetings from Poland
Daniel

### Example

```php
use Symfony\Component\Notifier\Message\SmsMessage;

$sms = new SmsMessage(
  // the phone number to send the SMS message to
  '+1411111111',
  // the message
  'A new login was detected!'
);
$sms->setFrom('+48 123 456 789');

$sentMessage = $texter->send($sms);
```

### PR DESCRIPTION
**Pull request by alamirault**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #45435 
| License       | MIT
| Doc PR        | symfony/symfony-docs#16701 [WIP]

This PR allow to define `From` in `SmsMessage`. When it defined, it override default `from` in transports.

```php
$sms = new SmsMessage(
  // the phone number to send the SMS message to
  '+1411111111',
  // the message
  'A new login was detected!',
  // [Optional] the From SMS message, it can also be a name
  '+1422222222',
);

// Can also be defined with setter
$sms->setFrom('+1422222222');

$sentMessage = $texter->send($sms);
```

### CODE REVIEW

File: src/Symfony/Component/Notifier/Message/SmsMessage.php [hunk 1]
```diff
@@ -93,6 +99,25 @@ public function getTransport(): ?string
         return $this->transport;
     }
 
+    /**
+     * @return $this
+     */
+    public function from(?string $from): static
```
Comment (alamirault) [inline src/Symfony/Component/Notifier/Message/SmsMessage.php:105]: Not know how to deal with setter (Need change in MessageInterface and introduce BC Break).

(Also not sure if setter is really needed)

File: src/Symfony/Component/Notifier/Message/SmsMessage.php [hunk 2]
```diff
@@ -93,6 +99,25 @@ public function getTransport(): ?string
...
+        if ('' === $from) {
+            throw new InvalidArgumentException(sprintf('"%s" needs a valid "from", it cannot be empty.', static::class));
+        }
+
+        $this->from = $from;
+
+        return $this;
+    }
+
+    public function getFrom(): ?string
```
Comment (alamirault) [inline src/Symfony/Component/Notifier/Message/SmsMessage.php]: Not know how to deal with new getter.
I think I have to create `SmsMessageInterface` but introduce BC Break

File: src/Symfony/Component/Notifier/Bridge/AllMySms/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+6.1
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/CHANGELOG.md]: 6.2 as 6.1 is now frozen.

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by alamirault** [comment]
Hope it's ok for you now

File: src/Symfony/Component/Notifier/Message/SmsMessage.php [hunk 4]
```diff
@@ -23,15 +23,21 @@ class SmsMessage implements MessageInterface
     private ?string $transport = null;
     private string $subject;
     private string $phone;
+    private ?string $from;
 
-    public function __construct(string $phone, string $subject)
+    public function __construct(string $phone, string $subject, string $from = null)
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Message/SmsMessage.php]:
```suggestion
    public function __construct(string $phone, string $subject, string $from = '')
```

File: src/Symfony/Component/Notifier/Message/SmsMessage.php [hunk 5]
```diff
@@ -23,15 +23,21 @@ class SmsMessage implements MessageInterface
     private ?string $transport = null;
     private string $subject;
     private string $phone;
+    private ?string $from;
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Message/SmsMessage.php]:
```suggestion
    private string $from;
```

File: src/Symfony/Component/Notifier/Message/SmsMessage.php [hunk 6]
```diff
@@ -23,15 +23,21 @@ class SmsMessage implements MessageInterface
...
 
-    public function __construct(string $phone, string $subject)
+    public function __construct(string $phone, string $subject, string $from = null)
     {
         if ('' === $phone) {
             throw new InvalidArgumentException(sprintf('"%s" needs a phone number, it cannot be empty.', __CLASS__));
         }
 
+        if ('' === $from) {
+            throw new InvalidArgumentException(sprintf('"%s" needs a valid "from", it cannot be empty.', __CLASS__));
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Message/SmsMessage.php]: I would remove this check and keep the empty string as "no from defined".

File: src/Symfony/Component/Notifier/Message/SmsMessage.php (hunk 1 above)
Comment (fabpot) [inline src/Symfony/Component/Notifier/Message/SmsMessage.php]:
```suggestion
    public function from(string $from): static
```

File: src/Symfony/Component/Notifier/Message/SmsMessage.php [hunk 7]
```diff
@@ -93,6 +99,25 @@ public function getTransport(): ?string
         return $this->transport;
     }
 
+    /**
+     * @return $this
+     */
+    public function from(?string $from): static
+    {
+        if ('' === $from) {
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Message/SmsMessage.php]: To be removed

File: src/Symfony/Component/Notifier/Message/SmsMessage.php (hunk 2 above)
Comment (fabpot) [inline src/Symfony/Component/Notifier/Message/SmsMessage.php]:
```suggestion
    public function getFrom(): string
```

File: src/Symfony/Component/Notifier/Bridge/AllMySms/CHANGELOG.md [hunk 8]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+6.2
+---
+
+* Use `SmsMessage->from` when is defined
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/CHANGELOG.md]:
```suggestion
* Use `SmsMessage->from` when defined
```

**Review by fabpot — CHANGES_REQUESTED** [review]
Can you also rebase on current 6.2 to get rid of the merge commit?

File: src/Symfony/Component/Notifier/Bridge/Smsapi/CHANGELOG.md [hunk 9]
```diff
@@ -1,11 +1,15 @@
...
 
+6.2
+---
+* Use `SmsMessage->from` when is defined
+
 6.1
 ---
 
- * Added `fast` option to the DSN that allows sending message with the highest priority that ensures the quickest possible time of delivery
- * Use `SmsMessage->from` when is defined
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Smsapi/CHANGELOG.md]: Please revert this change

File: src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php [hunk 10]
```diff
@@ -58,6 +58,7 @@ protected function doSend(MessageInterface $message): SentMessage
             throw new UnsupportedMessageTypeException(__CLASS__, SmsMessage::class, $message);
         }
 
+        /** @var SmsMessage $message */
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php:61]: I think only the Notifier Changeling and the Changeling of FreeMobile Bridge should be adjusted. All other bridges does not change and therefore doesn't need a changeling entry. 🤷‍♂️

File: src/Symfony/Component/Notifier/Message/SmsMessage.php (hunk 4 above)
Comment (alamirault) [reply src/Symfony/Component/Notifier/Message/SmsMessage.php]: @fabpot 

I'm not sure to understand why is it better ?

In transports we will have to do `$from = '' !== $message->getFrom() ? $message->getFrom() : $this->from;` instead of `$from = $message->getFrom() ?? $this->from;` Seems to be less clear

File: src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php (hunk 10 above)
Comment (alamirault) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php:61]: I think you reviewed only latest commits: there are code changes in all transports not only FreeMobile Bridge

Comment (OskarStark) [reply src/Symfony/Component/Notifier/Bridge/FreeMobile/FreeMobileTransport.php:61]: Indeed, thanks for the hint

File: src/Symfony/Component/Notifier/Message/SmsMessage.php (hunk 4 above)
Comment (fabpot) [reply src/Symfony/Component/Notifier/Message/SmsMessage.php]: I don't see the need to have `null` in addition to an empty string (an empty string is not a valid from, so we need to care of this anyway).

`$from = $message->getFrom() ? $message->getFrom() : $this->from` is enough.

**Comment by alamirault** [comment]
@fabpot @OskarStark  I rebased on current 6.2 and removed nullability of "from"

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @alamirault.
